language/move-prover/src/prover_task_runner.rs

```diff
@@ -32,20 +36,36 @@ pub trait ProverTask {
     fn init(&mut self, num_instances: usize) -> Vec<Self::TaskId>;
 
     /// Run the task with task_id. This function will be called from one of the worker threads.
-    async fn run(&mut self, task_id: Self::TaskId) -> Self::TaskResult;
+    async fn run(&mut self, task_id: Self::TaskId, sem: Arc<Semaphore>) -> Self::TaskResult;
+
+    /// Returns whether the task result is considered successful.
+    fn is_success(&self, task_result: &Self::TaskResult) -> bool;
 }
 
 pub struct ProverTaskRunner();
 
 impl ProverTaskRunner {
     /// Run `num_instances` instances of the prover `task` and returns the task id
     /// as well as the result of the fastest running instance.
-    pub fn run_tasks<T>(mut task: T, num_instances: usize) -> (T::TaskId, T::TaskResult)
+    pub fn run_tasks<T>(
+        mut task: T,
+        num_instances: usize,
+        sequential: bool,
+    ) -> (T::TaskId, T::TaskResult)
     where
         T: ProverTask + Clone + Send + 'static,
     {
-        let rt = Runtime::new().unwrap();
-
+        let rt = tokio::runtime::Builder::new()
+            .core_threads(2)
+            .threaded_scheduler()
+            .enable_all()
+            .build()
+            .unwrap();
+        let sem = if sequential {
+            Arc::new(Semaphore::new(1))
+        } else {
+            Arc::new(Semaphore::new(num_instances))
```

Comment:
Are you sure you need Arc around a semaphore? The type seems to implement the Sync trait, so it is save to pass between threads directly.

Reply:
Semaphore does implement Sync, but it doesn't implement Copy or Clone trait which is needed for us to pass the semaphore in at each iteration of the loop. If we pass the semaphore directly, Rust compiler complains that it has been moved in the last iteration of the loop.